### PR TITLE
[VIM-7221] Replaces AFNetworking's reachability manager

### PR DESCRIPTION
**Pull request by rpassis**
#### Ticket

[VIM-7221](https://vimean.atlassian.net/browse/VIM-7221)

#### Pull Request Checklist

- [X] Resolved any merge conflicts
- [X] No build errors or warnings are introduced
- [X] New files are written in Swift
- [X] New classes contain license headers
- [X] New classes have Documentation
- [X] New public methods have Documentation

#### Issue Summary

As part of the project to remove our dependency on AFNetworking, there are some key components that we will need to replace, one of each is the AFReachabilityManager class.

The goal of this PR is to introduce the replacement for this class that does not depend on third party libraries, and is written in Swift.

#### Implementation Summary

There are several implementations of `Reachability` available online, and our solution is heavily based in one of them - the basis of our `NetworkReachabilityManager` was taken from [Alamofire](https://raw.githubusercontent.com/Alamofire/Alamofire/master/Source/NetworkReachabilityManager.swift) and modified to our needs.

Most of the implementation of this class is to be kept internal to the framework, and we have refactored `VimeoReachability` into `VimeoReachabilityProvider`, and exposing the basic functionality required by consumers of the API through the `ReachabilityManagingType` protocol.

### CODE REVIEW

**Comment by cameo-github** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">SwiftLint not installed, download from https://github.com/realm/SwiftLint</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 135 tests, with 0 failures (0 unexpected) in 1.096 (1.262) seconds</td>
    </tr>
  </tbody>
</table>

See build details on [CircleCI](https://circleci.com/gh/vimeo/VimeoNetworking/1256)

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: Sources/Shared/VimeoReachability.swift [hunk 1]
```diff
@@ -1,38 +0,0 @@
-//
```
Comment (bryantjustin) [inline Sources/Shared/VimeoReachability.swift:1]: Good riddance. 👋

File: Sources/Shared/Reachability/NetworkReachabilityManager.swift [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+//  VimeoNetworking
+//
+//  Copyright © 2019 Vimeo. All rights reserved.
+//
+
+import Foundation
+import SystemConfiguration
+
+/// The protocol indicating the capabilities of a `ReachabilityManagingType` type
+public protocol ReachabilityManagingType {
```
Comment (bryantjustin) [inline Sources/Shared/Reachability/NetworkReachabilityManager.swift]: Is there a reason to prefer `ReachabilityManagingType` over simply `ReachabilityManaging`?

File: Sources/Shared/Reachability/NetworkReachabilityManager.swift [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    
+    /// `DispatchQueue` on which listeners will be called.
+    private var listenerQueue: DispatchQueue?
+    
+    /// `SCNetworkReachability` instance providing notifications.
+    private let reachability: SCNetworkReachability
+    
+    /// Storage for the previous status.
+    /// TODO: Ensure read/write access to this is synchronized [RDPA, 08/29/2019]
+    private var previousStatus: NetworkReachabilityStatus?
```
Comment (bryantjustin) [inline Sources/Shared/Reachability/NetworkReachabilityManager.swift:78]: Ahh this is why you were asking for the synchronization mechanism.

File: Sources/Shared/VimeoClient.swift [hunk 4]
```diff
@@ -89,6 +89,8 @@ final public class VimeoClient {
         /// response cache handles all memory and disk caching of response dictionaries
     private let responseCache = ResponseCache()
     
+    private var reachabilityMonitor: ReachabilityManagingType?
```
Comment (bryantjustin) [inline Sources/Shared/VimeoClient.swift]: Why the change in name here to `rechabilityMonitor`?

File: Sources/Shared/VimeoClient.swift [hunk 5]
```diff
@@ -108,17 +110,32 @@ final public class VimeoClient {
     /// - Parameters:
     ///   - appConfiguration: Your application's configuration
     ///   - configureSessionManagerBlock: a block to configure the session manager
-    convenience public init(appConfiguration: AppConfiguration, configureSessionManagerBlock: ConfigureSessionManagerBlock?) {
-        self.init(appConfiguration: appConfiguration, sessionManager: VimeoSessionManager.defaultSessionManager(appConfiguration: appConfiguration, configureSessionManagerBlock: configureSessionManagerBlock))
+    convenience public init(
+        appConfiguration: AppConfiguration,
+        configureSessionManagerBlock: ConfigureSessionManagerBlock? = nil,
+        reachabilityMonitor: ReachabilityManagingType? = VimeoReachabilityProvider.reachabilityManager
```
Comment (bryantjustin) [inline Sources/Shared/VimeoClient.swift]: Should we maybe switch the order of the parameters so trailing-closure syntax could be taken advantage of?

File: Sources/Shared/VimeoClient.swift [hunk 6]
```diff
@@ -159,7 +176,11 @@ final public class VimeoClient {
      
      - returns: a `RequestToken` for the in-flight request
      */
-    public func request<ModelType>(_ request: Request<ModelType>, completionQueue: DispatchQueue = DispatchQueue.main, completion: @escaping ResultCompletion<Response<ModelType>, NSError>.T) -> RequestToken {
+    public func request<ModelType>(
+        _ request: Request<ModelType>,
+        completionQueue: DispatchQueue = DispatchQueue.main,
```
Comment (bryantjustin) [inline Sources/Shared/VimeoClient.swift]: Nit: thoughts on just doing:
```suggestion
        completionQueue: DispatchQueue = .main,
```

File: Sources/Shared/VimeoClient.swift [hunk 7]
```diff
@@ -401,30 +438,38 @@ extension VimeoClient {
...
     /// client specific app configurations at start time.
     ///
     /// - Parameters:
     ///   - appConfiguration: An AppConfiguration instance
     ///   - configureSessionManagerBlock: a block to configure the session manager
-    public static func configureSharedClient(withAppConfiguration appConfiguration: AppConfiguration, configureSessionManagerBlock: ConfigureSessionManagerBlock?) {
+    public static func configureSharedClient(
+        withAppConfiguration appConfiguration: AppConfiguration,
+        configureSessionManagerBlock: ConfigureSessionManagerBlock? = nil,
+        reachabilityMonitor: ReachabilityManagingType? = VimeoReachabilityProvider.reachabilityManager
```
Comment (bryantjustin) [inline Sources/Shared/VimeoClient.swift]: Another spot where we could swap parameter orders to allow trailing-closure syntax at the invocation site.

File: Tests/Shared/Reachability/NetworkReachabilityStatusTests.swift [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            .reachable,
+            .connectionRequired,
+            .connectionOnTraffic,
+            .interventionRequired
+        ])
+        
+        // Then status should be not reachable
+        XCTAssertEqual(status, .notReachable)
+    }
+
```
Comment (bryantjustin) [inline Tests/Shared/Reachability/NetworkReachabilityStatusTests.swift]: Nit: extraneous spaces here

File: Tests/Shared/Reachability/NetworkingReachabilityManagerTests.swift [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+        manager?.startListening { status in
+            networkReachabilityStatus = status
+            expectation.fulfill()
+        }
+        waitForExpectations(timeout: 1, handler: nil)
+
+        // Then
+        XCTAssertEqual(networkReachabilityStatus, .reachable(.ethernetOrWiFi))
+    }
+
```
Comment (bryantjustin) [inline Tests/Shared/Reachability/NetworkingReachabilityManagerTests.swift]: Nit: extra space.

File: Tests/Shared/Reachability/NetworkingReachabilityManagerTests.swift [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+//  Created by Rogerio de Paula Assis on 8/29/19.
+//  Copyright © 2019 Vimeo. All rights reserved.
+//
+
+import XCTest
+@testable import VimeoNetworking
+
+class NetworkingReachabilityManagerTests: XCTestCase {
+
+    func testThatManagerCanBeInitializedFromHost() {
```
Comment (bryantjustin) [inline Tests/Shared/Reachability/NetworkingReachabilityManagerTests.swift]: The test methods here don't seem to follow our format in the style [guide](https://github.vimeows.com/MobileApps/Style-Guides/tree/master/Apple#naming-unit-tests).

**Review by bryantjustin — CHANGES_REQUESTED** [review]
A few small comments/suggestions. Looks good for the most part.

**Review by bryantjustin — APPROVED** [review]
LGTM.

**Pull request merged**